internal/crypto/ptls/common.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2024 the Pinniped contributors. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package ptls
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"slices"
+	"strings"
+	"sync/atomic"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"go.pinniped.dev/internal/plog"
+)
+
+// allowedCiphersForTLSOneDotTwo will only contain ciphers that meet the following criteria:
+// 1. They are secure
+// 2. They are returned by tls.CipherSuites
+// 3. They are returned by cipherSuitesForDefault or cipherSuitesForDefaultLDAP
+// This is atomic so that it can not be set and read at the same time.
+//
+//nolint:gochecknoglobals // This needs to be global because it will be set at application startup from configuration values
+var allowedCiphersForTLSOneDotTwo atomic.Value
+
+type SetAllowedCiphers func([]string) error
+
+// SetUserConfiguredCiphersForTLSOneDotTwo allows configuration/setup components to constrain the allowed TLS ciphers
+// for TLS1.2.
+func SetUserConfiguredCiphersForTLSOneDotTwo(userConfiguredCiphersForTLSOneDotTwo []string) error {
+	plog.Info("setting user-configured allowed ciphers for TLS 1.2", "userConfiguredAllowedCipherSuites", userConfiguredCiphersForTLSOneDotTwo)
+
+	validatedUserConfiguredAllowedCipherSuites, err := validateAllowedCiphers(allHardcodedAllowedCipherSuites(), userConfiguredCiphersForTLSOneDotTwo)
+	if err != nil {
+		return err
+	}
+	allowedCiphersForTLSOneDotTwo.Store(validatedUserConfiguredAllowedCipherSuites)
+	return nil
+}
+
+// getUserConfiguredCiphersAllowList returns the user-configured list of allowed ciphers for TLS1.2.
+// It is not exported so that it is only available to this package.
+func getUserConfiguredCiphersAllowList() []*tls.CipherSuite {
+	userConfiguredCiphersAllowList, ok := (allowedCiphersForTLSOneDotTwo.Load()).([]*tls.CipherSuite)
+	if ok {
+		return userConfiguredCiphersAllowList
+	}
+	return nil
+}
+
+// constrainCipherSuites returns the intersection of its parameters, as a list of cipher suite IDs.
+// If userConfiguredCiphersAllowList is empty, it will return all the hardcodedCipherSuites.
+func constrainCipherSuites(
+	hardcodedCipherSuites []*tls.CipherSuite,
+	userConfiguredAllowedCipherSuites []*tls.CipherSuite,
+) []uint16 {
+	allowedCipherSuiteIDs := sets.New[uint16]()
```

Comment:
Would something like this work as a performance optimization? The most common case will be that the user did not configure this new configuration option, so in that case can we avoid all this complex calculation cost?

```suggestion
        if Len(userConfiguredAllowedCipherSuites) == 0 {
          return hardcodedCipherSuites
        }
	allowedCipherSuiteIDs := sets.New[uint16]()
```

Reply:
Done.